library/core/src/ptr/mut_ptr.rs

```diff
@@ -1634,16 +1755,123 @@ impl<T: ?Sized> *mut T {
+    ///     // At compiletime, we know for sure that the pointer isn't aligned to 8.
+    ///     assert!(!ptr.is_aligned_to(8));
+    ///     assert!(!ptr.wrapping_add(1).is_aligned_to(8));
+    /// };
+    /// ```
+    ///
+    /// Due to this behavior, it is possible that a runtime pointer derived from a compiletime
+    /// pointer is aligned, even if the compiletime pointer wasn't aligned.
+    ///
+    #[cfg_attr(bootstrap, doc = "```ignore")]
+    #[cfg_attr(not(bootstrap), doc = "```")]
+    /// #![feature(pointer_is_aligned)]
+    /// #![feature(const_pointer_is_aligned)]
+    ///
+    /// // On some platforms, the alignment of i32 is less than 4.
+    /// #[repr(align(4))]
+    /// struct AlignedI32(i32);
+    ///
+    /// // At compiletime, neither `COMPTIME_PTR` nor `COMPTIME_PTR + 1` is aligned.
+    /// // Also, note that mutable references are not allowed in the final value of constants.
+    /// const COMPTIME_PTR: *mut AlignedI32 = (&AlignedI32(42) as *const AlignedI32).cast_mut();
+    /// const _: () = assert!(!COMPTIME_PTR.is_aligned_to(8));
+    /// const _: () = assert!(!COMPTIME_PTR.wrapping_add(1).is_aligned_to(8));
+    ///
+    /// // At runtime, either `runtime_ptr` or `runtime_ptr + 1` is aligned.
+    /// let runtime_ptr = COMPTIME_PTR;
+    /// assert_ne!(
+    ///     runtime_ptr.is_aligned_to(8),
+    ///     runtime_ptr.wrapping_add(1).is_aligned_to(8),
+    /// );
+    /// ```
+    ///
+    /// If a pointer is created from a fixed address, this function behaves the same during
+    /// runtime and compiletime.
+    ///
+    #[cfg_attr(bootstrap, doc = "```ignore")]
+    #[cfg_attr(not(bootstrap), doc = "```")]
+    /// #![feature(pointer_is_aligned)]
+    /// #![feature(const_pointer_is_aligned)]
+    ///
+    /// const _: () = {
+    ///     let ptr = 40 as *mut u8;
+    ///     assert!(ptr.is_aligned_to(1));
+    ///     assert!(ptr.is_aligned_to(2));
+    ///     assert!(ptr.is_aligned_to(4));
+    ///     assert!(ptr.is_aligned_to(8));
+    ///     assert!(!ptr.is_aligned_to(16));
+    /// };
+    /// ```
+    ///
+    /// [tracking issue]: https://github.com/rust-lang/rust/issues/104203
     #[must_use]
     #[inline]
     #[unstable(feature = "pointer_is_aligned", issue = "96284")]
-    pub fn is_aligned_to(self, align: usize) -> bool {
+    #[rustc_const_unstable(feature = "const_pointer_is_aligned", issue = "104203")]
+    pub const fn is_aligned_to(self, align: usize) -> bool {
         if !align.is_power_of_two() {
             panic!("is_aligned_to: align is not a power-of-two");
         }
 
-        // Cast is needed for `T: !Sized`
-        self.cast::<u8>().addr() & align - 1 == 0
+        // We can't use the address of `self` in a `const fn`, so we use `align_offset` instead.
+        // The cast to `()` is used to
+        //   1. deal with fat pointers; and
+        //   2. ensure that `align_offset` doesn't actually try to compute an offset.
+        self.cast::<()>().align_offset(align) == 0
```

Comment:
Sadly this caused a regression in Miri: https://github.com/rust-lang/miri/issues/2682

Reply:
While the immediate issue is fixed, it's still somewhat strange that `is_aligned` would change behavior with Miri's symbolic alignment mode... but maybe it makes sense, it is consistent with `align_to`, anyway. We'll have to watch out for other similar regressions. If too many bugreports come in we'll have to find another solution.

Reply:
The last line of this doc test is also failing with `-Zmiri-symbolic-alignment-check`
https://github.com/rust-lang/rust/blob/c9c017dfb55e375800c3e424311939a7ba3d4deb/library/core/src/ptr/const_ptr.rs#L1492-L1511

Maybe we should partially revert daccb8c11a464040acd744729231429dcb2e4d4b to put the `const_eval_select` back?

It might also make sense to redefine `-Zmiri-symbolic-alignment-check` as "runtime alignment behaves like const eval alignment", because i think that is what it currently does after this PR and https://github.com/rust-lang/miri/pull/2683.

Reply:
We are not running libcore tests with symbolic alignment so I guess I didn't notice this.

> It might also make sense to redefine -Zmiri-symbolic-alignment-check as "runtime alignment behaves like const eval alignment", because i think that is what it currently does after this PR and https://github.com/rust-lang/miri/pull/2683.

I guess that makes sense. Are you proposing just a docs change or also an implementation change?

Reply:
We could do some deduplication between the [ctfe impl](https://github.com/rust-lang/rust/blob/master/compiler/rustc_const_eval/src/const_eval/machine.rs#L216-L269) and [miri impl](https://github.com/rust-lang/rust/blob/master/src/tools/miri/src/shims/mod.rs#L67-L112) of `align_offset`, but the implementation looks functionally identical to me, so this would mostly be a docs change.

Currently, the docs look like this:
> `-Zmiri-symbolic-alignment-check` makes the alignment check more strict. By default, alignment is checked by casting the pointer to an integer, and making sure that is a multiple of the alignment. This can lead to cases where a program passes the alignment check by pure chance, because things "happened to be" sufficiently aligned -- there is no UB in this execution but there would be UB in others. To avoid such cases, the symbolic alignment check only takes into account the requested alignment of the relevant allocation, and the offset into that allocation. This avoids missing such bugs, but it also incurs some false positives when the code does manual integer arithmetic to ensure alignment. (The standard library align_to method works fine in both modes; under symbolic alignment it only fills the middle slice when the allocation guarantees sufficient alignment.)

From this it actually seems pretty clear to me that new behavior for `is_aligned` with `-Zmiri-symbolic-alignment-check` is correct and the old one was wrong. Also, it seems weird to me that the docs don't mention `align_offset` at all when literally all this flag does is change the behavior of `align_offset`.

Maybe we could just change the last sentence in parentheses to something like
> (This changes the runtime behavior of alignment-related standard library functions like `is_aligned`, `align_offset`, or `align_to` to match the compiletime behavior. For example, `align_to` only fills the middle slice when the allocation guarantees sufficient alignment.)

Reply:
>  Also, it seems weird to me that the docs don't mention align_offset at all when literally all this flag does is change the behavior of align_offset.

It does more. It's core feature is to toggle a flag in the interpreter that affects how alignment checking works. Adjusting `align_offset` is just a side thing that we also do to keep more code working in this mode.